core/server/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -309,8 +309,8 @@ public CreatePathResult createPath(AlluxioURI path, CreatePathOptions options)
         // to the non-persisted Inodes of traversalResult.
         traversalResult.getNonPersisted().add(lastInode);
         toPersistDirectories.add(lastInode);
-      } else if (!(options instanceof CreateDirectoryOptions && ((CreateDirectoryOptions) options)
-          .isAllowExists())) {
+      } else if (!(lastInode.isDirectory() && options instanceof CreateDirectoryOptions
```

Comment:
I pushed it down one of the conjuction, but kept the other one since tying the `instanceof` with casting the options object makes sense to me ... does that work for you?